### ISSUE TITLE
I am trying to make DTLS work with our customized BIO (block io with timeout). Found that if  I manually drop packets, the handshake will fail.

### ISSUE THREAD

**Issue by YaoDing** [openssl/openssl#10908]
````
    Openssl version: 1.1.1a
````
  
I used BIO_meth_new to create our customized BIO and set function pointers with our blocked IO methods(with timeout).
On client side I do SSL_connect(Thread 1), and on server side I do SSL_accept(Thread 2).
I have another two threads keep calling DTLSv1_get_timeout to see if timeout expires and if it does I will do DTLSv1_handle_timeout (I am running client and server in a single process, my customized bio is really just memory copy/move). 
n my case, for example, if I drop the first packet on one side when it reads something(since I am using our own BIO method, in our bio_read function, I can just drop packets), and even though I see dtlsv1_handle_timeout returns 1 (which means it's resending packet), the handshake still failed.
Please advise. Thank you!

**Comment by mattcaswell** [comment]
Is it possible to get a packet trace of what is actually being sent between the client/server?

**Comment by YaoDing** [comment]
@mattcaswell Thank you for your reply.
In the case I dropped the first packet received on both sides,
below is the messages exchanged in between:
````
client is sending: 16,fe,ff,00,00,00,00,00,00,00,00,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

client is sending: 16,fe,ff,00,00,00,00,00,00,00,01,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

server is sending: 16,fe,ff,00,00,00,00,00,00,00,00,00,15,03,00,00,09,00,00,00,00,00,00,00,09,fe,ff,06,63,6f,6f,6b,69,65

client is sending: 16,fe,ff,00,00,00,00,00,00,00,02,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

client is sending: 16,fe,ff,00,00,00,00,00,00,00,03,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

client is sending: 16,fe,ff,00,00,00,00,00,00,00,04,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

client is sending: 16,fe,ff,00,00,00,00,00,00,00,05,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

client is sending: 16,fe,ff,00,00,00,00,00,00,00,06,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

client is sending: 16,fe,ff,00,00,00,00,00,00,00,07,00,c6,01,00,00,ba,00,00,00,00,00,00,00,ba,fe,fd,a7,d8,83,2b,b3,ac,db,8a,da,51,bc,bf,b5,27,8a,ce,8d,0c,fb,92,3d,e7,96,12,81,ec,56,46,4d,fc,89,1d,00,00,00,38,c0,2c,c0,30,00,9f,cc,a9,cc,a8,cc,aa,c0,2b,c0,2f,00,9e,c0,24,c0,28,00,6b,c0,23,c0,27,00,67,c0,0a,c0,14,00,39,c0,09,c0,13,00,33,00,9d,00,9c,00,3d,00,3c,00,35,00,2f,00,ff,01,00,00,58,00,0b,00,04,03,00,01,02,00,0a,00,0c,00,0a,00,1d,00,17,00,1e,00,19,00,18,00,23,00,00,00,16,00,00,00,17,00,00,00,0d,00,30,00,2e,04,03,05,03,06,03,08,07,08,08,08,09,08,0a,08,0b,08,04,08,05,08,06,04,01,05,01,06,01,03,03,02,03,03,01,02,01,03,02,02,02,04,02,05,02,06,02

````

**Comment by YaoDing** [comment]
Another thing is after I added printing messages in bio_write and if I drop the 2nd packets on both sides, sometimes I saw crashes on calling dtls1_retransmit_message (Line 1020 in statem_dtls.c) since flag(item->data) is some random value.

**Comment by mattcaswell** [comment]
So what you have here is:

````
client is sending: 16,fe,ff,00,00,00,00,00,00,00,00,00,c6,01...
````
The first 13 bytes are the record header. 16 indicates this is a handshake message. fe, ff indicates the protocol version. The next 8 bytes (all zeroes) are the record epoch and sequence number. 00, c6 is the length of the record data. The next byte is the interesting one: 01 indicates a ClientHello. 03 indicates a HelloVerifyRequest.

So this means what you have is:

Client sends ClientHello
Client sends ClientHello
Server sends HelloVerifyRequest
Client sends ClientHello
Client sends ClientHello
Client sends ClientHello
Client sends ClientHello
Client sends ClientHello
Client sends ClientHello

The first couple of messages are easy enough to explain. The client starts out by sending a ClientHello and it gets dropped. It then retransmits it. The server has been configured to require a cookie and sends back a HelloVerifyRequest. This also gets dropped. The server doesn't retransmit the HelloVerifyRequest and is waiting for a ClientHello with the cookie in it. The client meanwhile is stuck retransmitting ClientHellos that don't have the cookie that are being ignored by the server. IIRC HelloVerifyRequest messages aren't ever put into the retransmit queue (because the server is supposed to be operating statelessly until such time as we have a verified ClientHello).

If you're going to require cookies on the server side then you should be using DTLSv1_listen():

https://www.openssl.org/docs/man1.1.1/man3/DTLSv1_listen.html

**Comment by YaoDing** [comment]
Thank you, @mattcaswell 
Actually I used DTLSV1_listen before and it worked on linux but not on windows 10. So I removed it and magically still worked on both platforms. Thought maybe it's okay to not call it.
Now I added it back and yes as I saw before, it was stuck in DTLSv1_listen (on windows only). 
However, there's a magic again that if I link against libssl_static.lib and libcrypto_static.lib instead of libssl.lib and libcrypto.lib, then problem's solved. DTLSv1_listen worked perfectly with/without packets dropped.
Can you share more information for the difference between libxxx.lib and libxxx_static.lib?

**Comment by YaoDing** [comment]
@mattcaswell , I guess I figured it out. The thing is the *.lib I am using is not actually a static library, it still requires dll. And even I did not see any dlls in my bin folder (Which why I was thinking I was using a static library which contains all implementations) and then at run time, my example app went to some other folders to load dlls whose version I don't know could be not supporting dtls. I am gonna close this issue. Thanks!
